src/lsm/tree.zig

```diff
@@ -701,28 +787,76 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             assert(tree.compaction_callback != null);
 
             // compact_done() is called after all compact_tick()'s complete.
-            // This function can be triggered asynchronously or by compact_cpu() below.
             tree.compaction_io_pending -= 1;
             if (tree.compaction_io_pending == 0) tree.compact_done();
         }
 
+        /// Called at the end of each compaction tick.
         fn compact_done(tree: *Tree) void {
             assert(tree.compaction_io_pending == 0);
             assert(tree.compaction_callback != null);
 
-            // Will be true if there's any Compactions still processing.
-            var still_compacting = false;
+            // At the end of every beat:
+            // - ensure mutable table can be flushed to immutable table.
+            defer assert(tree.table_mutable.can_commit_batch(tree.options.commit_count_max));
 
-            // Reset the immutable table Compaction once it finishes.
-            // Also clear any invisible tables it created during compaction.
             const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
             const even_levels = compaction_beat < half_measure_beat_count;
-            if (even_levels) {
-                assert(tree.compaction_table_immutable.status == .idle);
-            } else {
+            const finished_even_levels = compaction_beat == half_measure_beat_count - 1;
+            const finished_odd_levels = compaction_beat == config.lsm_batch_multiple - 1;
+            if (!finished_even_levels and !finished_odd_levels) {
+                const callback = tree.compaction_callback.?;
+                tree.compaction_callback = null;
+                callback(tree);
+                return;
```

Comment:
Could this code be moved into a `compact_finish()` function which is also called by `manifest_checkpoint_callback()`?